p2p/pex/addrbook.go

```diff
@@ -405,89 +406,10 @@ func (a *addrBook) GetSelectionWithBias(biasTowardsNewAddrs int) []*p2p.NetAddre
-
-	oldBucketToAddrsMap := make(map[int]map[string]struct{})
-	var oldIndex int
-	newBucketToAddrsMap := make(map[int]map[string]struct{})
-	var newIndex int
-
-	// initialize counters used to count old and new added addresses.
-	// len(oldBucketToAddrsMap) cannot be used as multiple addresses can endup in the same bucket.
-	var oldAddressesAdded int
-	var newAddressesAdded int
-
 	// number of new addresses that, if possible, should be in the beginning of the selection
-	numRequiredNewAdd := percentageOfNum(biasTowardsNewAddrs, numAddresses)
-
-	selectionIndex := 0
-ADDRS_LOOP:
-	for selectionIndex < numAddresses {
-		// biasedTowardsOldAddrs indicates if the selection can switch to old addresses
-		biasedTowardsOldAddrs := selectionIndex >= numRequiredNewAdd
-		// An old addresses is selected if:
-		// - the bias is for old and old addressees are still available or,
-		// - there are no new addresses or all new addresses have been selected.
-		// numAddresses <= a.nOld + a.nNew therefore it is guaranteed that there are enough
-		// addresses to fill the selection
-		pickFromOldBucket :=
-			(biasedTowardsOldAddrs && oldAddressesAdded < a.nOld) ||
-				a.nNew == 0 || newAddressesAdded >= a.nNew
-
-		bucket := make(map[string]*knownAddress)
-
-		// loop until we pick a random non-empty bucket
-		for len(bucket) == 0 {
-			if pickFromOldBucket {
-				oldIndex = a.rand.Intn(len(a.bucketsOld))
-				bucket = a.bucketsOld[oldIndex]
-			} else {
-				newIndex = a.rand.Intn(len(a.bucketsNew))
-				bucket = a.bucketsNew[newIndex]
-			}
-		}
-
-		// pick a random index
-		randIndex := a.rand.Intn(len(bucket))
-
-		// loop over the map to return that index
-		var selectedAddr *p2p.NetAddress
-		for _, ka := range bucket {
-			if randIndex == 0 {
-				selectedAddr = ka.Addr
-				break
-			}
-			randIndex--
-		}
-
-		// if we have selected the address before, restart the loop
-		// otherwise, record it and continue
-		if pickFromOldBucket {
-			if addrsMap, ok := oldBucketToAddrsMap[oldIndex]; ok {
-				if _, ok = addrsMap[selectedAddr.String()]; ok {
-					continue ADDRS_LOOP
-				}
-			} else {
-				oldBucketToAddrsMap[oldIndex] = make(map[string]struct{})
-			}
-			oldBucketToAddrsMap[oldIndex][selectedAddr.String()] = struct{}{}
-			oldAddressesAdded++
-		} else {
-			if addrsMap, ok := newBucketToAddrsMap[newIndex]; ok {
-				if _, ok = addrsMap[selectedAddr.String()]; ok {
-					continue ADDRS_LOOP
-				}
-			} else {
-				newBucketToAddrsMap[newIndex] = make(map[string]struct{})
-			}
-			newBucketToAddrsMap[newIndex][selectedAddr.String()] = struct{}{}
-			newAddressesAdded++
-		}
-
-		selection[selectionIndex] = selectedAddr
-		selectionIndex++
-	}
-
+	numRequiredNewAdd := cmn.MaxInt(percentageOfNum(biasTowardsNewAddrs, numAddresses), numAddresses-a.nOld)
+	selection := a.randomPickAddresses(bucketTypeNew, numRequiredNewAdd)
+	selection = append(selection, a.randomPickAddresses(bucketTypeOld, numAddresses-len(selection))...)
```

Comment:
Before this change the selection was:
 - percentageOfNum(biasTowardsNewAddrs, numAddresses) -> new addresses
 - old addresses, and
 - if there are not enough old addresses to fill the selection, more new addresses are added at the end of the selection.
With this change the old addresses (a.k.a. good addresses) are pushed back to the end of the selection. 
This means that the pex peer that receives the selection will process the good addresses last, perhaps delaying connections to good peers. I'm not sure if this is a real issue though.
On the other hand it's a simple change to make it back the way it was.

Reply:
> This means that the pex peer that receives the selection will process the good addresses last, perhaps delaying connections to good peers. I'm not sure if this is a real issue though.

Not an issue

Reply:
Reason being because the peer who receives them doesn't know which are old and which are new, only the sending peer knows from their own perspective. So from the receiver's side, they are all just added to the addrbook as "new" until they prove themselves.